### PR TITLE
Add a new version of ContentionEnd event

### PR DESCRIPTION
**Pull request by gleocadie**
Add a new version of the ContenionEnd event which contains the duration of the contention.

#21524

### CODE REVIEW

**Comment by gleocadie** [comment]
cc @vancem

**Comment by gleocadie** [comment]
cc @noahfalk

**Comment by noahfalk** [comment]
I think @kouvel is the right person to look at it. It seemed reasonable to me, though I did wonder if we should be guarding the work to track the duration with a conditional so that we don't do it when the event isn't going to be emitted. My guess is that the two QPC + QueryPerformanceFrequency is 200-250 cycles of additional delay and significant portion is being incurred while holding the contended lock. I'm fine with whatever call @kouvel wants to make.

**Comment by gleocadie** [comment]
Fine by me, I will had the check before emitting the event.
Thanks for the feedback.
@kouvel , I'll let you know when it's done thx.

File: src/vm/syncblk.cpp [hunk 1]
```diff
@@ -2554,6 +2554,17 @@ inline void LogContention()
...
 
+LONGLONG ComputeElapsedTimeInNanosecond(LARGE_INTEGER startTime, LARGE_INTEGER endTime)
+{
+    static LARGE_INTEGER freq;
+    if (freq.QuadPart == 0)
+        QueryPerformanceFrequency(&freq);
+
+    const int NsPerSecond = 1000 * 1000 * 1000;
+    LONGLONG elapsedTicks = endTime.QuadPart - startTime.QuadPart;
+    return (elapsedTicks * NsPerSecond) / freq.QuadPart;
```
Comment (kouvel) [inline src/vm/syncblk.cpp]: It would probably make sense to compute and return a double value without int truncation

File: src/vm/syncblk.cpp [hunk 2]
```diff
@@ -2554,6 +2554,17 @@ inline void LogContention()
 #define LogContention()
 #endif
 
+LONGLONG ComputeElapsedTimeInNanosecond(LARGE_INTEGER startTime, LARGE_INTEGER endTime)
```
Comment (kouvel) [inline src/vm/syncblk.cpp]: Maybe startTicks / endTicks instead of startTime / endTime?

File: src/vm/syncblk.cpp [hunk 3]
```diff
@@ -2571,8 +2582,16 @@ BOOL AwareLock::EnterEpilogHelper(Thread* pCurThread, INT32 timeOut)
     // the object associated with this lock.
     _ASSERTE(pCurThread->PreemptiveGCDisabled());
 
-    // Fire a contention start event for a managed contention
-    FireEtwContentionStart_V1(ETW::ContentionLog::ContentionStructs::ManagedContention, GetClrInstanceId());
+    BOOLEAN IsContentionKeywordEnabled = ETW_TRACING_CATEGORY_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_Context, TRACE_LEVEL_INFORMATION, CLR_CONTENTION_KEYWORD);
+    LARGE_INTEGER startTime = {0};
```
Comment (kouvel) [inline src/vm/syncblk.cpp]: Maybe startTicks instead of startTime?

File: src/vm/syncblk.cpp [hunk 4]
```diff
@@ -2704,8 +2723,17 @@ BOOL AwareLock::EnterEpilogHelper(Thread* pCurThread, INT32 timeOut)
     GCPROTECT_END();
     DecrementTransientPrecious();
 
-    // Fire a contention end event for a managed contention
-    FireEtwContentionStop(ETW::ContentionLog::ContentionStructs::ManagedContention, GetClrInstanceId());
+    if (IsContentionKeywordEnabled)
+    {
+        LARGE_INTEGER endTime;
```
Comment (kouvel) [inline src/vm/syncblk.cpp]: Maybe endTicks instead of endTime?

File: src/vm/syncblk.cpp (hunk 1 above)
Comment (gleocadie) [reply src/vm/syncblk.cpp]: you are right.

File: src/vm/syncblk.cpp (hunk 2 above)
Comment (gleocadie) [reply src/vm/syncblk.cpp]: done

File: src/vm/syncblk.cpp (hunk 3 above)
Comment (gleocadie) [reply src/vm/syncblk.cpp]: done

File: src/vm/syncblk.cpp (hunk 4 above)
Comment (gleocadie) [reply src/vm/syncblk.cpp]: done

**Comment by kouvel** [comment]
Thanks! Failures are not related to the change.

**Comment by gleocadie** [comment]
Thanks

**Pull request merged**